appendix-backends.Rmd

```diff
@@ -0,0 +1,98 @@
+|----------------|----------------------------------------------------------------------------------------------|-----------------------------------|-----------------------------------|---------------------------------------------------------------------------------------------|
+| bootDiskSizeGB | n/a                                                                                          | n/a                               | n/a                               | Request a disk of this size to boot the Docker image (useful for very large Docker images)  |
+| cpu            | Reserve this many cores (acts as a maximum)                                                  | n/a                               |                                   | Request a minimum of this many cores (scales with memory)                                   |
+| disks          | n/a                                                                                          | n/a                               |                                   | Request this much disk size - soft requirement, if not specified, will request 10 GB        |
+| docker         | Run the task in this Docker image                                                            | Run the task in this Docker image | Run the task in this Docker image | Run the task in this Docker image                                                           |
+| maxRetries     |                                                                                              |                                   |                                   | If task errors, retry this number of times (additive with preemptible)                      |
+| memory         | Maximum amount of memory to use                                                              | n/a                               |                                   | Minimum amount of memory to use (scales with CPU)                                           |
+| preemptible    | n/a                                                                                          | n/a                               | n/a                               | Attempt running on a preemptible instance this many times, then switch to a non-preemptible |
+| walltime       | How much  [walltime](https://en.wikipedia.org/wiki/Elapsed_real_time)  to request for a task | n/a                               | n/a                               | n/a                                                                                         |
+
+::: {.notice data-latex="warning"}
+Although Cromwell-on-local and miniwdl-on-local ignore some runtime attributes, Cromwell (on default settings) appears to be more likely to cause problems when running scattered tasks due to mishandling system resources. This can include locking up Docker Engine for the entire system! See the Cromwell-on-local section below for more details.
```

Comment:
In brief, the reason is that if I throw 100 samples at a workflow that scatters per sample, and have miniwdl run it, it work without a problem, but if I do the same with Cromwell it will explode.

It's a tricky thing to talk about because:
* Almost all of my experience with this issue is in the context of running on a local compute, which is common for testing workflows but atypical for running big ones.
* It's not just a matter of "this will work on miniwdl but fail on Cromwell," the Cromwell breakage actually affects the Docker daemon. When this happens, it is difficult to diagnose unless you know exactly what you're looking for. If you're running on a shared compute resource, this could seriously break things for other people. It's sort of a "don't assemble a genome on the head node if you don't want a curt email from IT and your colleagues" situation.
* I've never seen this happen with miniwdl, but it is theoretically possible, so I can't speak in absolutes.
* It's not clear exactly why this is happening. It appears that Cromwell and miniwdl both attempt to spin up as many concurrent Docker images as they think they can manage at the same time, but miniwdl may be more conservative in its guessing. It could also have to do with Cromwell not supporting the `memory` runtime attribute as universally as miniwdl does.
* The issue has been reported to Cromwell's devs, but it's not prioritized as it doesn't happen on the backends they focus the most on (eg, Terra and Azure).

Reply:
Tried to reword this in the latest commits, let me know if you think I threaded the needle properly this time.